tests/lib/nested.sh

```diff
@@ -0,0 +1,210 @@
+
+    if [ "${ENABLE_TPM:-false}" = "true" ]; then
+        if [ -d "/tmp/qtpm" ]; then
+            rm -rf /tmp/qtpm
+        fi
+
+        # remove the swtpm
+        # TODO: we could just remove/reset the swtpm instead of removing the snap 
+        # wholesale
+        snap remove swtpm-mvo
+    fi
+
+    # delete the image file
+    rm -rf "${IMAGE_FILE}"
+}
+
+print_nested_status(){
+    SVC_NAME="nested-vm-$(systemd-escape "${SPREAD_JOB:-unknown}")"
+    systemctl status "${SVC_NAME}" || true
+    journalctl -u "${SVC_NAME}" || true
+}
+
+start_nested_core_vm_unit(){
+    # copy the image file to create a new one to use
+    # TODO: maybe create a snapshot qcow image instead?
+    mkdir -p "${WORK_DIR}"
+    cp "${SETUPDIR}/pc.img" "${IMAGE_FILE}"
+
+    # use only 2G of RAM for qemu-nested
+    if [ "${SPREAD_BACKEND}" = "google-nested" ]; then
+        PARAM_MEM="-m 4096"
+        PARAM_SMP="-smp 2"
+    elif [ "${SPREAD_BACKEND}" = "lxd-nested" ]; then
+        PARAM_MEM="-m 4096"
+        PARAM_SMP="-smp 2"
+    elif [ "${SPREAD_BACKEND}" = "qemu-nested" ]; then
+        PARAM_MEM="-m 2048"
+        PARAM_SMP="-smp 1"
+    else
+        echo "unknown spread backend ${SPREAD_BACKEND}"
+        exit 1
+    fi
+
+    PARAM_DISPLAY="-nographic"
+    PARAM_NETWORK="-net nic,model=virtio -net user,hostfwd=tcp::${SSH_PORT}-:22"
+    # TODO: do we need monitor port still?
+    PARAM_MONITOR="-monitor tcp:127.0.0.1:${MON_PORT},server,nowait"
+    PARAM_RANDOM="-object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0"
+    PARAM_CPU=""
+    PARAM_TRACE="-d cpu_reset"
+    PARAM_LOG="-D ${WORK_DIR}/qemu.log"
+    PARAM_SERIAL="-serial file:${WORK_DIR}/serial.log"
+    PARAM_TPM=""
+
+    ATTR_KVM=""
+    if [ "$ENABLE_KVM" = "true" ]; then
+        ATTR_KVM=",accel=kvm"
+        # CPU can be defined just when kvm is enabled
+        PARAM_CPU="-cpu host"
+        # Increase the number of cpus used once the issue related to kvm and ovmf is fixed
+        # https://bugs.launchpad.net/ubuntu/+source/kvm/+bug/1872803
+        PARAM_SMP="-smp 1"
+    fi
+
+    # TODO: enable ms key booting for i.e. nightly edge jobs ?
+    OVMF_CODE=""
+    OVMF_VARS=""
+    if [ "${ENABLE_SECURE_BOOT:-false}" = "true" ]; then
+        OVMF_CODE=".secboot"
+    fi
+    if [ "${ENABLE_OVMF_SNAKEOIL:-false}" = "true" ]; then
+        OVMF_VARS=".snakeoil"
+    fi
+
+    mkdir -p "${WORK_DIR}/image/"
+    cp -f "/usr/share/OVMF/OVMF_VARS${OVMF_VARS}.fd" "${WORK_DIR}/image/OVMF_VARS${OVMF_VARS}.fd"
+    PARAM_BIOS="-drive file=/usr/share/OVMF/OVMF_CODE${OVMF_CODE}.fd,if=pflash,format=raw,unit=0,readonly -drive file=${WORK_DIR}/image/OVMF_VARS${OVMF_VARS}.fd,if=pflash,format=raw"
+    PARAM_MACHINE="-machine q35${ATTR_KVM} -global ICH9-LPC.disable_s3=1"
+
+    if [ "${ENABLE_TPM:-false}" = "true" ]; then
+        TPMSOCK_PATH="/var/snap/swtpm-mvo/current/swtpm-sock"
+        if [ "${SPREAD_BACKEND}" = "lxd-nested" ]; then
```

Comment:
It is not obvious to me why we use a snap in some cases and not others. This is the kind of thing that needs a comment.